File: circular_city/shuffle_actions.py
"""Shuffle event actions so display order varies (circular is not always first)."""

from __future__ import annotations


def _hash_seed(seed: str | int) -> int:
    s = 0
    for ch in str(seed):
        s = (s * 31 + ord(ch)) & 0xFFFFFFFF
    return s or 1


def shuffle_actions(actions: list[dict], seed: str | int) -> list[dict]:
    if not actions:
        return []
    arr = list(actions)
    s = _hash_seed(seed)
    for i in range(len(arr) - 1, 0, -1):
        s = (s * 1103515245 + 12345) & 0x7FFFFFFF
        j = s % (i + 1)
        arr[i], arr[j] = arr[j], arr[i]
    return [
        {"action": action, "display_letter": chr(65 + index)}
        for index, action in enumerate(arr)
    ]


def shuffle_seed_for_event(event: dict | None, city_id: str | None, round_num: int | None) -> str:
    return f"{event.get('id') if event else 'ev'}:{city_id or 'city'}:{round_num or 0}"

File: src/game/shuffleActions.js
/** Shuffle event actions so display order varies (circular choices are not always first). */

function hashSeed(seed) {
  let s = 0;
  const str = String(seed);
  for (let i = 0; i < str.length; i++) {
    s = (s * 31 + str.charCodeAt(i)) >>> 0;
  }
  return s || 1;
}

export function shuffleActions(actions, seed) {
  if (!actions?.length) return [];
  const arr = [...actions];
  let s = hashSeed(seed);
  for (let i = arr.length - 1; i > 0; i--) {
    s = (s * 1103515245 + 12345) & 0x7fffffff;
    const j = s % (i + 1);
    [arr[i], arr[j]] = [arr[j], arr[i]];
  }
  return arr.map((action, index) => ({
    action,
    displayLetter: String.fromCharCode(65 + index),
  }));
}

export function shuffleSeedForEvent(event, cityId, round) {
  return `${event?.id ?? 'ev'}:${cityId ?? 'city'}:${round ?? 0}`;
}

/** Shuffle insight (justify) options; returns new correctIndex for display order. */
export function shuffleJustifyOptions(justify, seed) {
  if (!justify?.options?.length) {
    return { ...justify, options: [], correctIndex: 0 };
  }
  const indexed = justify.options.map((text, i) => ({ text, i }));
  let s = hashSeed(`${seed}:justify`);
  for (let i = indexed.length - 1; i > 0; i--) {
    s = (s * 1103515245 + 12345) & 0x7fffffff;
    const j = s % (i + 1);
    [indexed[i], indexed[j]] = [indexed[j], indexed[i]];
  }
  return {
    question: justify.question,
    conceptTag: justify.conceptTag,
    options: indexed.map((x) => x.text),
    correctIndex: indexed.findIndex((x) => x.i === justify.correctIndex),
  };
}
